### PR TITLE
add conditions to build api

### PR DESCRIPTION
**Pull request by bparees**
implementation of https://jira.coreos.com/browse/DEVEXP-95

api updates: https://github.com/openshift/api/pull/532
client-go updates(using temporary bump from bparees fork): https://github.com/openshift/client-go/pull/128
openshift-apiserver updates(using temporary bump from bparees fork): https://github.com/openshift/openshift-apiserver/pull/52
origin(e2e test, using temporary bump from bparees fork) updates: https://github.com/openshift/origin/pull/24254

### CODE REVIEW

**Comment by bparees** [comment]
@adambkaplan Once this PR goes green, we can merge https://github.com/openshift/api/pull/532 and https://github.com/openshift/client-go/pull/128, then i'll come back and update this PR to use a proper bump.  Then once it merges, we can retest/merge https://github.com/openshift/origin/pull/24254

File: pkg/build/apiserver/buildgenerator/generator.go [hunk 1]
```diff
@@ -536,6 +536,12 @@ func (g *BuildGenerator) generateBuildFromConfig(ctx context.Context, bc *buildv
 		},
 		Status: buildv1.BuildStatus{
 			Phase: buildv1.BuildPhaseNew,
+			Conditions: []buildv1.BuildCondition{{
+				Type:               buildv1.BuildConditionType(buildv1.BuildPhaseNew),
+				Status:             corev1.ConditionTrue,
+				LastUpdateTime:     metav1.Now(),
+				LastTransitionTime: metav1.Now(),
```
Comment (adambkaplan) [inline pkg/build/apiserver/buildgenerator/generator.go]: nit - use same object for `LastUpdateTime` and `LastTransitionTime`

File: pkg/build/apiserver/buildgenerator/generator.go [hunk 2]
```diff
@@ -879,7 +885,13 @@ func generateBuildFromBuild(build *buildv1.Build, buildConfig *buildv1.BuildConf
 			OwnerReferences: buildCopy.ObjectMeta.OwnerReferences,
 		},
 		Status: buildv1.BuildStatus{
-			Phase:  buildv1.BuildPhaseNew,
+			Phase: buildv1.BuildPhaseNew,
+			Conditions: []buildv1.BuildCondition{{
+				Type:               buildv1.BuildConditionType(buildv1.BuildPhaseNew),
+				Status:             corev1.ConditionTrue,
+				LastUpdateTime:     metav1.Now(),
+				LastTransitionTime: metav1.Now(),
```
Comment (adambkaplan) [inline pkg/build/apiserver/buildgenerator/generator.go]: nit - use same object for timestamps

File: pkg/build/apiserver/registry/build/strategy.go [hunk 3]
```diff
@@ -37,12 +38,58 @@ func (strategy) AllowUnconditionalUpdate() bool {
...
+				}
+				build.Status.Conditions[i].LastUpdateTime = metav1.Now()
+				build.Status.Conditions[i].Reason = string(build.Status.Reason)
+				build.Status.Conditions[i].Message = build.Status.Message
+			}
+		} else {
+			if c.Status != kapi.ConditionFalse {
+				build.Status.Conditions[i].Status = kapi.ConditionFalse
+				build.Status.Conditions[i].LastTransitionTime = metav1.Now()
+				build.Status.Conditions[i].LastUpdateTime = metav1.Now()
```
Comment (adambkaplan) [inline pkg/build/apiserver/registry/build/strategy.go]: nit - use same object for timestamps

File: pkg/build/apiserver/registry/build/strategy.go [hunk 4]
```diff
@@ -37,12 +38,58 @@ func (strategy) AllowUnconditionalUpdate() bool {
...
+				build.Status.Conditions[i].Message = ""
+			}
+		}
+	}
+	if !found {
+		condition := buildapi.BuildCondition{
+			Type:               buildapi.BuildConditionType(build.Status.Phase),
+			Status:             kapi.ConditionTrue,
+			LastUpdateTime:     metav1.Now(),
+			LastTransitionTime: metav1.Now(),
```
Comment (adambkaplan) [inline pkg/build/apiserver/registry/build/strategy.go]: nit - use same object for timestamps

**Comment by bparees** [comment]
@adambkaplan updated.  if you can lgtm https://github.com/openshift/api/pull/532 i'll start the process of merging/bumping across the repos.

**Comment by bparees** [comment]
/hold cancel

@adambkaplan this is ready for re-review/merge.

**Comment by bparees** [comment]
/retest

@adambkaplan ptal

**Pull request merged**